File: bench/suite/src/st/binary-trees/pascal/binarytrees.pas
{ binary-trees (Benchmarks Game algorithm), single-threaded, plain Free Pascal.
  Every node is individually allocated with New and released with Dispose. }
program binarytrees;
{$mode objfpc}

type
  PNode = ^TNode;
  TNode = record
    left, right: PNode;
  end;

function Make(depth: longint): PNode;
begin
  New(Result);
  if depth > 0 then
  begin
    Result^.left := Make(depth - 1);
    Result^.right := Make(depth - 1);
  end
  else
  begin
    Result^.left := nil;
    Result^.right := nil;
  end;
end;

function Check(t: PNode): longint;
begin
  if t^.left <> nil then
    Result := 1 + Check(t^.left) + Check(t^.right)
  else
    Result := 1;
end;

procedure Free(t: PNode);
begin
  if t^.left <> nil then
  begin
    Free(t^.left);
    Free(t^.right);
  end;
  Dispose(t);
end;

const
  MIN_DEPTH = 4;

var
  n, maxDepth, stretchDepth, depth, iterations, i, c, code: longint;
  t, longLived: PNode;
begin
  n := 10;
  if ParamCount >= 1 then Val(ParamStr(1), n, code);
  maxDepth := n;
  if maxDepth < MIN_DEPTH + 2 then maxDepth := MIN_DEPTH + 2;
  stretchDepth := maxDepth + 1;

  t := Make(stretchDepth);
  WriteLn('stretch tree of depth ', stretchDepth, #9' check: ', Check(t));
  Free(t);

  longLived := Make(maxDepth);
  depth := MIN_DEPTH;
  while depth <= maxDepth do
  begin
    iterations := 1 shl (maxDepth - depth + MIN_DEPTH);
    c := 0;
    for i := 1 to iterations do
    begin
      t := Make(depth);
      c := c + Check(t);
      Free(t);
    end;
    WriteLn(iterations, #9' trees of depth ', depth, #9' check: ', c);
    Inc(depth, 2);
  end;
  WriteLn('long lived tree of depth ', maxDepth, #9' check: ', Check(longLived));
  Free(longLived);
end.

File: bench/suite/src/st/binary-trees/fortran/binarytrees.f90
! binary-trees (Benchmarks Game algorithm), single-threaded, plain Fortran.
! Every node is individually allocated and deallocated (no pool).
module trees
  implicit none
  type node
    type(node), pointer :: left => null()
    type(node), pointer :: right => null()
  end type
contains
  recursive function make(depth) result(t)
    integer, intent(in) :: depth
    type(node), pointer :: t
    allocate(t)
    if (depth > 0) then
      t%left => make(depth - 1)
      t%right => make(depth - 1)
    end if
  end function

  recursive integer function check(t) result(c)
    type(node), pointer :: t
    if (associated(t%left)) then
      c = 1 + check(t%left) + check(t%right)
    else
      c = 1
    end if
  end function

  recursive subroutine free(t)
    type(node), pointer :: t
    if (associated(t%left)) then
      call free(t%left)
      call free(t%right)
    end if
    deallocate(t)
  end subroutine
end module trees

program binarytrees
  use trees
  implicit none
  integer, parameter :: min_depth = 4
  character, parameter :: tab = achar(9)
  integer :: n, max_depth, stretch_depth, depth, iterations, i, c
  type(node), pointer :: t, long_lived
  character(len=32) :: arg

  n = 10
  if (command_argument_count() >= 1) then
    call get_command_argument(1, arg)
    read (arg, *) n
  end if
  max_depth = max(min_depth + 2, n)
  stretch_depth = max_depth + 1

  t => make(stretch_depth)
  write (*, '(2(a,i0))') 'stretch tree of depth ', stretch_depth, tab//' check: ', check(t)
  call free(t)

  long_lived => make(max_depth)

  do depth = min_depth, max_depth, 2
    iterations = ishft(1, max_depth - depth + min_depth)
    c = 0
    do i = 1, iterations
      t => make(depth)
      c = c + check(t)
      call free(t)
    end do
    write (*, '(2(i0,a),i0)') iterations, tab//' trees of depth ', depth, tab//' check: ', c
  end do

  write (*, '(2(a,i0))') 'long lived tree of depth ', max_depth, tab//' check: ', check(long_lived)
  call free(long_lived)
end program binarytrees
